Block visibility based on screen size: add backend block support

ramonjd opened the pull request:
Related to https://github.com/WordPress/gutenberg/issues/72502

Backport Core PR:

- https://github.com/WordPress/wordpress-develop/pull/10629

## What?

Implements backend support for responsive block visibility, allowing blocks to be hidden on specific viewport breakpoints (mobile, tablet, desktop) using CSS media queries.

Functionality is gated behind the `gutenberg-hide-blocks-based-on-screen-size` experiment flag. The existing boolean blockVisibility: false behavior works without the experiment enabled.

Clientside state is happening over in:

- https://github.com/WordPress/gutenberg/pull/74025

## How?
This PR borrows approaches from #73735 and #73888 to implement the backend that:

- generates media queries based on block metadata
- applies the classname and enqueues the CSS via the style engine

## Testing


Unit tests:


`npm run test:unit:php:base -- --filter=WP_Block_Supports_Block_Visibility_Test`

To test manually, enable the experiment:

<img width="998" height="90" alt="Screenshot 2025-12-15 at 10 53 52 am" src="https://github.com/user-attachments/assets/ec7052f1-3758-4f2c-8ad1-93ddd50fd0c5" />

Then copy and paste this into your post editor, publish and view the frontend. 

```html


<!-- wp:paragraph {"metadata":{"blockVisibility":{"mobile":false}}} -->
<p>🔴 Hidden on MOBILE (≤599px) - You should see this on tablet and desktop only</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"metadata":{"blockVisibility":{"tablet":false}}} -->
<p>🟡 Hidden on TABLET (600-959px) - You should see this on mobile and desktop only</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"metadata":{"blockVisibility":{"desktop":false}}} -->
<p>🟢 Hidden on DESKTOP (≥960px) - You should see this on mobile and tablet only</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"metadata":{"blockVisibility":{"mobile":false,"desktop":false}}} -->
<p>🟣 Hidden on MOBILE and DESKTOP - You should only see this on tablet (600-959px)</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"metadata":{"blockVisibility":{"mobile":false,"tablet":false}}} -->
<p>🔵 Hidden on MOBILE and TABLET - You should only see this on desktop (≥960px)</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"metadata":{"blockVisibility":{"tablet":false,"desktop":false}}} -->
<p>🟠 Hidden on TABLET and DESKTOP - You should only see this on mobile (≤599px)</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"metadata":{"blockVisibility":{"mobile":true,"tablet":true,"desktop":true}}} -->
<p>✅ Visible on ALL breakpoints - You should always see this</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"metadata":{"blockVisibility":false}} -->
<p>❌ Completely HIDDEN (boolean false) - You should never see this</p>
<!-- /wp:paragraph -->
```

https://github.com/user-attachments/assets/45f86b73-f8f8-4763-a1eb-2303fda96105

Make sure to test with the experiment off as well. The last item should never show with the experiment on and off.

ramonjd committed:
* Update block visibility rendering to support responsive visibility based on screen size.
* Introduce a new function to add 'display' to safe CSS properties for responsive styles.
* Add comprehensive unit tests to validate block visibility behavior across different scenarios, including responsive breakpoints and experiment toggling.

ramonjd committed:
Remove deprecated function for adding 'display' to safe CSS properties and corresponding unit tests. This streamlines block visibility support by eliminating unused code related to responsive visibility experiments.

ramonjd committed:
Add compatibility for KSES in WordPress 7.0

- Introduce a new file for KSES compatibility, adding 'display' to the list of safe CSS properties to support responsive visibility.
- Update unit tests to verify the correct generation of CSS styles, including 'display:none' for hidden blocks based on screen size.

ramonjd committed:
Extract breakpoint definitions into a separate array.

PrathameshBhagat365 reviewed (commented):
Sanitize generated class names: use sanitize_html_class() (or sanitize_key()) when building 'wp-block-hidden-...'.
Restrict/validate breakpoints: only honor known breakpoint keys (array_intersect with defined breakpoints) to avoid mismatched counts and unexpected classes.
All-hidden check: compare hidden breakpoints against actual breakpoint keys (use intersection) before returning ''.
WP_HTML_Tag_Processor fallback: add a safe fallback if the class isn't available (careful regex or DOMDocument) so adding the class can't fail.
KSES shim loading: guard or document the compat/kses.php include (WP version check or comment) and add a unit test that toggles the experiment and asserts 'display' is added/removed from safe_style_css.
Style engine / !important: confirm the engine accepts '!important' values; prefer explicit API for importance or ensure the generated CSS is not stripped/altered.
Tests: add test for unknown breakpoint keys being ignored and for the WP_HTML_Tag_Processor fallback (if implemented).

ramonjd committed:
Support by sanitizing breakpoint names and ignoring unknown breakpoints. Update unit tests to validate behavior for responsive visibility scenarios, ensuring only known breakpoints are processed and that valid class names are generated.

ramonjd committed:
add comment about breakpoint

ramonjd committed:
Add block visibility support updates and new changelog entry

- Introduce a new changelog file for pull request #10629.
- Update comments in block visibility support to clarify future JSON format centralization.
- Refactor unit test to reflect Core

mtias commented on lib/block-supports/block-visibility.php:
```diff
@@ -6,7 +6,7 @@
 
 /**
- * Render nothing if the block is hidden.
+ * Render nothing if the block is hidden, or add responsive visibility styles.
```
I wonder if we should call this "viewport" not "responsive" in general.

ramonjd replied:
"viewport", in my mind, does come with less semantic baggage. 👍🏻

ramonjd committed:
Update block visibility support to use viewport terminology

ramonjd committed:
- Update comments to clarify behavior when blocks are hidden on all breakpoints.
- Simplify CSS generation by removing unnecessary sanitization of breakpoint names in the visibility class.
- Remove outdated unit test for CSS generation related to display:none, ensuring tests focus on current functionality.

ramonjd committed:
Add unit test for block visibility support when all breakpoints are hidden

- Introduce a new test to ensure that block content is empty when all viewport breakpoints (mobile, tablet, desktop) are set to hidden.
- Update existing test assertions for clarity regarding empty visibility objects.

tellthemachines commented on lib/block-supports/block-visibility.php:
```diff
@@ -19,10 +19,109 @@ function gutenberg_render_block_visibility_support( $block_content, $block ) {
+				'max' => '599px',
+			),
+			'tablet'  => array(
+				'min' => '600px',
```
If we're thinking about what shape this might have as a theme.json setting, defining both min and max breakpoints feels redundant, because the min of the next breakpoint is always the max of the previous one plus one px. 

On the other hand, once we allow defining custom breakpoints we should support other units such as `em` and `rem`, which are commonly used to build media queries responsive to browser font size settings. I suppose for those cases we could do something like `calc( 30rem - 1px)` [calc in media queries ](https://caniuse.com/mdn-css_at-rules_media_calc) seems to have solid support nowadays.

ramonjd replied:
>  the min of the next breakpoint is always the max of the previous one plus one px.

I appreciate the feedback! 

True, I agree that defining both min and max is redundant for a simple sequential breakpoint system, but can we guarantee that any list is going to be linearly sequential and in px units? No.

I wanted the data structure to be flexible enough to support more complex conditions and, as you said, units, without boxing us into a rigid JSON or any accompanying string parsing logic. 

Here's where I was going with `min` and `max`:

- No parsing required: values are used exactly as written
- No order assumptions: each breakpoint is self-contained
- Explicit and clear: anyone reading this knows exactly what ranges are active
- Works with any CSS value: "37.4375rem", "calc(50vw - 20px)", anything
- Supports gaps and overlaps

It can also be extended, e.g., 

```json
    "landscape-large": {
      "min": "1024px",
      "orientation": "landscape"
    },
```


That said, I'm up for anything, especially a design that makes the default case simpler while still allowing flexibility. Happy to explore alternative structures if you have thoughts on how to handle these cases!

tellthemachines replied:
> can we guarantee that any list is going to be linearly sequential and in px units?

We're building the API, so we write the rules 😄  If we say that the breakpoints have to be sequential, users will know to define them correctly.

The thing is, I can't imagine any plausible scenario where it would be advantageous to define breakpoints out of sequence. What problem are we trying to solve here?

The other thing we should consider is whether we wish to support a potentially unlimited number of breakpoints, e.g.:

<img width="268" height="354" alt="Screenshot 2025-12-18 at 9 06 17 am" src="https://github.com/user-attachments/assets/c72f894a-53fc-436b-87b4-0daa07c4d3ba" />

Or whether we want to constrain the breakpoints that can be defined to the existing mobile, tablet and desktop.

ramonjd replied:
> We're building the API, so we write the rules 😄 If we say that the breakpoints have to be sequential, users will know to define them correctly.
> I can't imagine any plausible scenario where it would be advantageous to define breakpoints out of sequence

I was thinking about users entering values in theme.json, and we know users will imagine and will do anything they want. 😆 

WordPress/Gutenberg has a history of needing to make breaking changes when initial assumptions prove wrong, that's why my instinct was towards flexibility. 

I mentioned gaps and overlaps and orientation above - they're contrived admittedly, and maybe not advantageous to 80% of folks 🤷🏻 but they're scenarios.

Is the concern preventing user error? 

I'd flip this question: what problem does enforcing sequential order solve that clear documentation and examples can't?

> The other thing we should consider is whether we wish to support a potentially unlimited number of breakpoints

I think this touches on the core issue. If we're considering unlimited breakpoints (which I think is right for extensibility), then enforcing rigid sequential rules becomes harder to maintain and validate.

tellthemachines replied:
My main concern is that we don't build something that becomes unmaintainable through being over-complicated, and that we _do_ build something consistent with the type of system we're offering with theme.json configuration: it's flexible, but it presents a sensible set of options to configure. It's not the whole CSS spec.

The sensible approach to responsive design is to work with the minimal amount of breakpoints that ensure elements reflow to fit any screen. Theme designers tend to be particularly good at CSS and layouts, so I doubt they'd have a need for unlimited breakpoints.

A restricted set of mobile, tablet, desktop breakpoints also ensures that themes can be changed and nothing breaks when certain blocks are hidden at certain breakpoints. Otherwise, how will we deal with translating between this theme's "movie screen" breakpoint and the other's "70 inch tv" one? Or are we just going to reset all hidden blocks if a user switches themes?

ramonjd replied:
> The sensible approach to responsive design is to work with the minimal amount of breakpoints that ensure elements reflow to fit any screen. Theme designers tend to be particularly good at CSS and layouts, so I doubt they'd have a need for unlimited breakpoint
> Otherwise, how will we deal with translating between this theme's "movie screen" breakpoint and the other's "70 inch tv" one? Or are we just going to reset all hidden blocks if a user switches themes?

Themes already have their own presets (colors, fontsizes etc) that don't translate across theme switches.  The system already handles this pattern - the CSS just isn't generated if you switch themes. I don't think theme switching creates any new problems that theme.json doesn't already solve for other presets.

I get your point that breakpoint property names could be standardized and portable across themes, I can imagine that it would be very helpful in addition to users being able to tweak their values.

To be honest, I found myself speculating a lot as well and didn't want to assume too much, which is why I leant towards flexibility.

I'm not pushing back, I just want to understand the concern better, and your solution so we can build something that help users. If you could illustrate what you had in mind, maybe we could propose something different.

E.g., a compromise might be the following. Start with defined presets, then later open up to "customBreakpoints"?

```json
{
  "settings": {
    "layout": {
      "breakpoints": {
        "mobile": { "max": "599px",  "label": "Mobile devices" },
        "tablet": { "min": "600px", "max": "959px", "label": "whateves" },
        "desktop": { "min": "960px", "label": "monster" }
      },
      "customBreakpoints": {
        "sidebar-collapsed": { "max": "300px" }
      }
    }
  }
}
```

Here's what I'm thinking:

1. Themes can customize the values but not add new breakpoint names
2. Solves theme switching: all themes understand mobile, tablet, desktop, flexibility without breaking theme portability
3. Themes can change the px/rem values whatever
4. Doesn't require sequential ordering enforcement and the overhead
5. Future iteration: optional theme-specific breakpoints - they don't apply on theme switch (with a warning/documentation)



Thanks for the discussion! We'll get there and it'll be great!

tellthemachines replied:
> I don't think theme switching creates any new problems that theme.json doesn't already solve for other presets.

Losing a colour or a font size won't change the actual content of the page though. With the show/hide behaviour, if you lose the breakpoints, suddenly all blocks on the page will be visible at every breakpoint, likely duplicating content. The breakpoints might also be used for responsive behaviour in layout blocks, so the potential breakage if they stop applying can get pretty bad. It would make it really hard to switch themes.

My ideal API would be:

```
"mediaBreakpoints": {
        "mobile": "600px",
        "tablet": "960px"
      }
```
With anything else we need calculated from those values. That will still allow using rem and so forth, we just need to structure the media query like: `min-width: calc( 45rem + 1px)` which has fine browser support nowadays. 

It's not clear that we'll need both min and max for any other use than the hiding of blocks, and I suspect it's a pretty niche use case even for that. (Seriously, why would you need to hide a block only on a tablet? Hardly anyone even uses tablets to browse the web.) For layout things you usually need to do stuff like "make this Columns block have two columns at 960px and one column at 600px" or even just "make it one col from 960px down". If the latter translates to 

```
@media (max-width: 600px) {
    flex-basis: 100% !important;
}
@media (min-width: 601px) and (max-width: 960px) {
    flex-basis: 100% !important;
}
```
we will rightly be accused of generating shitty code bloat 😂 

My concern with custom breakpoints is that once we allow them, there's no going back: we'll have to support them forever. Whereas allowing only mobile/tablet doesn't stop us from extending later. 

Also we should bear in mind that theme authors can always use additional CSS for really custom stuff.

ramonjd replied:
Thanks for the detailed response! I think we're getting closer to the core trade-offs.

> My concern with custom breakpoints is that once we allow them, there's no going back... Whereas allowing only mobile/tablet doesn't stop us from extending later.

Good point, and I agreed above with the "start with defined presets, then later open up to "customBreakpoints". I think we're on the same page there - we can always open the discussion up more widely for custom breakpoints and see if there's appetite/a way forward.

> That will still allow using rem and so forth, we just need to structure the media query like: min-width: calc( 45rem + 1px)

What sort of parsing do you expect to happen here? Do you think we should support things like `"calc(50vw - 20px)" > "calc(calc(50vw - 20px) + 1px)"`

> It's not clear that we'll need both min and max for any other use than the hiding of blocks, and I suspect it's a pretty niche use case even for that.

Isn't hiding blocks the core of the issue:

- https://github.com/WordPress/gutenberg/issues/72502

? 

What needs to change in the requirements?

>  Seriously, why would you need to hide a block only on a tablet? Hardly anyone even uses tablets to browse the web.)

Is that true?

Tablets often have unique UI constraints (landscape vs portrait, touch but larger screen), and if tablets are even 3-4% of global web traffic that's hundreds of millions of users. 

What's the argument for not hiding a block on a tablet exactly?

>  shitty code bloat

I'm not sure I understand your example, which has 2 x similar CSS declarations. Are you arguing against bad implementation, or against min/max structure ? We can have explicit min/max in the data and still generate optimized CSS, e.g., group like rules.

> Also we should bear in mind that theme authors can always use additional CSS for really custom stuff.

True, and this can be used an argument for any design tool we build. What's the goal here? 

I want to stress I'm not going to die on this hill, but I'm trying to arrive at something that fits theme.json's philosophy of "sensible defaults + customization". I'm thinking something simple for most users but doesn't paint us into a corner later. We cannot predict every use case, but we can support low-cost measures to increase support.

E.g., later if someone comes along and says "I need landscape-specific breakpoints" or "I need a breakpoint that only applies between 500-800px" we've got it covered, and don't have to worry about breaking changes.

> My ideal API would be:

On question of data structure, I think we're closer than we thought. 

A compromise could be data types. Forget custom breakpoints for the minute. 

1. Array > Sequential mode (simple, auto-calculated ranges)
2. Object > explicit mode

So there's no mixing ambiguity and we're flexible for later.

Not sure we even need `"label"` but I added it for the sake of the example:

### Sequential (Array)

```json
{
  "settings": {
    "layout": {
      "breakpoints": [
        { "slug": "mobile", "value": "37.4375rem", "label": "Nokia 3210" },
        { "slug": "tablet", "value": "59.9375rem", "label": "Nintendo Switch 2" }
      ]
    }
  }
}
```

- Array order defines sequence
- Values treated as max-width
 - System generates calc() for ranges, e.g., 
    - Mobile: @media (max-width: 37.4375rem)
    - Tablet: @media (min-width: calc(37.4375rem + 0.0625rem)) and (max-width: 59.9375rem)
    - Desktop (implicit): @media (min-width: calc(59.9375rem + 0.0625rem))

### Explicit Control (When needed) and can be implemented later

```json
{
  "settings": {
    "layout": {
      "breakpoints": {
        "mobile": {
          "max": "37.4375rem"
        },
        "tablet": {
          "min": "40rem",
          "max": "59.9375rem"
        },
        "desktop": {
          "label": "something",
          "min": "60rem"
        }
      }
    }
  }
}
```

- Uses values exactly as specified
- No sorting, no calculation
- Generates:
    - Mobile: @media (max-width: 37.4375rem)
    - Tablet: @media (min-width: 40rem) and (max-width: 59.9375rem)
    - Desktop: @media (min-width: 60rem)

ramonjd replied:
Another thing I forgot to add is, this discussion is pretty important, but I can't see any blockers right now for this PR while things are hardcoded. Does that sound right?

Kicking the can and all that 😄 

Great to have these issues in mind for later.

tellthemachines replied:
> What sort of parsing do you expect to happen here? Do you think we should support things like `"calc(50vw - 20px)" > "calc(calc(50vw - 20px) + 1px)"`

I'd rather we only support straightforward number + unit values; there _shouldn't_ be a need for anything else in media queries. Though nested calc functions are valid, so your example should work.

> Is that true?

[Stats](https://gs.statcounter.com/platform-market-share/desktop-mobile-tablet/worldwide) indicate it. Doesn't mean we shouldn't support it, but from the perspective of someone who is building a website, there's hardly a business case to optimise for it. What is increasingly important to optimise for is mobile, which has high usage in most regions and particularly high in developing countries. 

Anyway, the point I meant to make is that showing/hiding something on tablet only is likely a niche requirement. We should support it, but we shouldn't build our whole logic around that use case.

> We can have explicit min/max in the data and still generate optimized CSS, e.g., group like rules.

It's precisely the explicit min and max that I'm not on board with 😄 

It needlessly complicates the API. I shouldn't have to calculate a min breakpoint based on the max of the previous breakpoint plus 1 for things to work as I'd expect them to work. Forcing users to do that introduces scope for error in return for a possibility of defining arbitrary breakpoints (?) which is very unclear to me that there'll ever be a need for. 

We should start simple and complicate later if we really have to. 

> I can't see any blockers right now for this PR while things are hardcoded

Will we be able to change the shape of the breakpoints object that I asked the initial question about later? That's really the only point of contention I think.

ramonjd replied:
> Will we be able to change the shape of the breakpoints object that I asked the initial question about later?

Everything else aside, type-based rule detection would allow it - see Sequential (Array) in the comment above.

andrewserong replied:
Great discussion here. Without trying to complicate things, my gut instinct is:

* We should be able to support an ultra simple set of declarations (essentially the sequential approach) and:
* In the future, extend to support min / max / custom values or whatever feedback comes up if the sequential approach proves limiting

In practice, I imagine this would mean designing the API so that if we need to add `min` / `max` keys in the future we could? E.g. like we've done elsewhere, using a simple string value for now (i.e. possibly a `size` key to match other presets like font sizes and spacing sizes), but later on also support an object with keys for min / max etc?

ramonjd replied:
> We should be able to support an ultra simple set of declarations (essentially the sequential approach) and:
> In the future, extend to support min / max / custom values or whatever feedback comes up if the sequential approach proves limiting

I think that's where we landed? 

I was mainly worried (future angst admittedly) about a data structure that couldn't be extended. 

I think the array vs object approach could work both semantically and in practice seeing as arrays have built-in indexing/sequence.

I can update this PR to reflect that and we can see later how it works when value config comes up.

Thanks for the help and the discussion, folks!

andrewserong replied:
> I was mainly worried (future angst admittedly) about a data structure that couldn't be extended.

A natural worry to have, we've seen how adding features to `theme.json` winds up resulting in _heaps_ of feedback from folks trying to use things in unexpected ways!

> I can update this PR to reflect that and we can see later how it works when value config comes up.

Sounds good to me 👍

ramonjd replied:
Okay done. I appreciate all the back and forth. It helps, really. 😄

tellthemachines commented on phpunit/block-supports/block-visibility-test.php:
```diff
@@ -0,0 +1,398 @@
+		$this->assertSame( $block_content, $result );
+	}
+
+	public function test_block_visibility_support_without_experiment() {
```
Is it really worth adding this test? We'll have to remove it as soon as the experiment is stabilised, and unless it's a _very_ long running experiment the probability of accidental regressions is pretty low.

ramonjd replied:
It was just there for peace of mind really. We don't really need it, agree.

ramonjd replied:
Done!

tellthemachines reviewed (commented):
This is working well and generally LGTM, the only thing I'm not convinced about is the need for defining both max and min breakpoints (see below). This is assuming that we're shaping this in a way that will make it easy to build a theme.json setting on top of it.

ramonjd committed:
Remove unit test for block visibility support without the viewport visibility experiment.

ramonjd committed:
Refactor block visibility support to enhance breakpoint handling

- Update breakpoint definitions to include names and slugs for better clarity.
- Simplify media query generation logic for breakpoints, ensuring correct CSS output.
- Modify unit tests to validate the new structure and ensure accurate CSS generation for responsive visibility.

ramonjd commented on lib/block-supports/block-visibility.php:
```diff
@@ -19,10 +19,133 @@ function gutenberg_render_block_visibility_support( $block_content, $block ) {
+		 *
+		 * Breakpoints as array items are defined sequentially. The first item's size is the max value.
+		 * Each subsequent item's min is calc(previous size + 1px), and its size is the max.
+		 * The last item's min is calc(previous size + 1px) with no max.
```
```suggestion
		 * The last item's min is previous size with no max.
```

ramonjd committed:
Update lib/block-supports/block-visibility.php

ramonjd committed:
Yoda, go home

andrewserong commented on lib/block-supports/block-visibility.php:
```diff
@@ -19,10 +19,133 @@ function gutenberg_render_block_visibility_support( $block_content, $block ) {
+				$query_parts[] = '(max-width: ' . $size . ')';
+			} elseif ( count( $breakpoints ) - 1 === $index ) {
+				// Last item: min = calc(previous size + 1px), no max.
+				$query_parts[] = '(min-width: ' . $previous_size . ')';
```
Tiny nit: should this use the `calc` rule to add `1px` to the min-width, just like in the `else` block below?

ramonjd replied:
Oh yeah. That slipped through. Let me check it out. Thanks!

ramonjd replied:
It was originally with a calc(), but I reverted that for some unknown reason.

andrewserong commented on lib/block-supports/block-visibility.php:
```diff
@@ -19,10 +19,133 @@ function gutenberg_render_block_visibility_support( $block_content, $block ) {
+			return $block_content;
+		}
+
+		// If the block is hidden on all breakpoints, do not render the block.
```
Not for now necessarily, but eventually if we support customisation and only a single breakpoint (e.g. mobile and no others) then the presence of a single breakpoint might mean that we do still want to output it. As a single breakpoint would mean just a `max-width` rule so could still be useful.

ramonjd replied:
If I understand correctly, if a theme defines a single breakpoint, and hides a block on that breakpoint then the count check below would return `''`.

Is that what you mean? If so, another good point. Maybe I'll add a comment to remind us later.

andrewserong replied:
> Is that what you mean?

Yep! The expected result would be "hide on mobile" but instead it'd be treated as "hide always".

Again, nothing to action now, just thoughts for when we get to customisation!

andrewserong commented on lib/block-supports/block-visibility.php:
```diff
@@ -19,10 +19,133 @@ function gutenberg_render_block_visibility_support( $block_content, $block ) {
+		$css_rules = array();
+
+		foreach ( $hidden_on as $breakpoint ) {
+			$visibility_class = 'wp-block-hidden-' . $breakpoint;
```
For now it's fine because we have a controlled list of breakpoints, but we might want to wrap the `$breakpoint` in `sanitize_title` to ensure it gets the kebab-case treatment (and similarly below on line 141)

ramonjd replied:
Good point! I removed sanitization from https://github.com/WordPress/gutenberg/pull/73994/commits/6dd86daa05fa3df4c07d9e666f8cc2d769dfd9e5 as an early optimization. I think we could combine these two loops somehow however, or build the classlist so we don't have to loop through the breakpoints again below.

ramonjd replied:
I added a comment.

andrewserong commented on lib/compat/wordpress-7.0/kses.php:
```diff
@@ -0,0 +1,26 @@
+		return $attr;
+	}
+
+	if ( ! in_array( 'display', $attr, true ) ) {
+		$attr[] = 'display';
+	}
```
I can't quite remember the details but I seem to recall there being a little pushback a few years ago when trying to get `display` added in core to the list of safe CSS properties. Not a blocker for this PR, as this is guarded behind the experiment. But, if it does wind up being an issue getting it in core, then we might need a workaround for the style engine's parsing of the CSS rules.

In any case, good thing you've got a core PR up in https://github.com/WordPress/wordpress-develop/pull/10629 — I'm sure the discussion can continue there and we can continue to update this in Gutenberg as / when needed.

ramonjd replied:
I recall that too. ~I did some searching but couldn't find it.~ Found it, I think. 

If memory serves me correct, it was to prevent authors hiding site elements for spammy links:

https://github.com/WordPress/wordpress-develop/pull/2928#discussion_r912554761

I'm not convinced of the strength of this argument any more given that Google et. al punish such behaviour.

Any why should WordPress be the arbiter of what an author can and cannot do on their site, especially when it comes to a widely-used and inherently safe CSS property?

What do you reckon @peterwilsoncc? Merry Xmas!

ramonjd replied:
While I think `display: none` is fit for purpose here (we want to hide the element permanently and it is entirely non-essential to all users), the only other thing I can think of is using alternative CSS to "hide" the element?

E.g.,  https://www.a11y-collective.com/blog/visually-hidden/

andrewserong replied:
Yeah, I think `display: none` is the right fit for this support, too. On that other thread you previously made a good point that there's already lots of ways elements can be hidden with the existing list of CSS properties, even `margin-left: -99999px`, so I do think it's a good idea to revisit the discussion.

andrewserong reviewed (approved):
This is testing great for me both with and without the experiment active, and nice test coverage 👍

I left a few comments, but nothing major (or a blocker to merging) and mostly I'm getting excited about customisability so consider the comments thoughts rather than anything that needs changing just yet.

LGTM!

ramonjd committed:
- Update media query logic to use calc() for min-width values, ensuring accurate breakpoint handling.
- Refactor class name generation to accumulate visibility classes for multiple breakpoints.
- Add new unit tests for tablet and desktop breakpoints, validating CSS output and class attributes.
- Update existing tests for clarity and consistency in visibility class assertions.

ramonjd commented:
Thanks for testing @andrewserong 

I've addressed the `calc` issue, updated the tests and added comments for TODO if ever these breakpoints are customizable. 

I'll let the tests run, then merge. We're behind the experiment!

ramonjd committed:
updated comment

ramonjd committed:
Remove unused method for disabling viewport visibility experiment in block visibility tests
